### PR TITLE
fix: (Test Generation) Better Error Messages and Variable Names

### LINKED ISSUE #3708: Fail on test generation if no tests are actually created
### Summary

If no actual tests are generated, when passing in the `generate-tests` flag, the output of Dafny should be non-zero, even if it generated syntactically correct output.

### Background and Motivation

It can occur, especially when `assume` or axioms are used, that *no* methods are called in the generated tests.

This can be confusing, since the generated Dafny is still valid, but no methods annotated with `{:test}` are generated, so the whole module is a no-op.

### Proposed Feature

I propose that when no methods annotated with `{:test}` occur in the output (or more generally when no methods are generated) that Dafny exit with a non-zero exit code, ideally with an informative error message, perhaps

```
Error: No tests were generated, because no code points could be proven reachable in method foo (do you have a false assumption in the method?)
```

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by Dargones**
Fixes #3708, https://github.com/dafny-lang/dafny/issues/1477

This PR improves test generation based on recent feedback by adding clearer error messages and variable names that are based on the given variable's type. Full changelog:

- Terminate test generation with a dedicated error if no test methods are generated.
- Terminate test generation with a dedicated error if the target method cannot be found (previously the program would crush).
- Improve explanation of cases in which a test is not generated due to verification being successfull or timing out.
- Name generated variables in accordance with their type (e.g. int0 and bool0 as opposed to v0 and v1).
- Fix several typos.
- [Update] Made sure DafnyOption instance is passed from the driver to the test generation entry point
- [Update] Made Test Generation backwards compatible with Z3 4.8.5 in the same way as the LanguageServer does it
- [Update] Fix an issue with Arrow type parsing in counterexamples that seems to affect test generation (the issue concerns a corner case where a type is an instance of UserDefinedType but AsArrowType() does not return null and in fact the type is an arrow type)

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Review by atomb — CHANGES_REQUESTED** [review]
This is a nice improvement! There will need to be some changes to resolve conflicts with a recent PR, though.

File: Source/DafnyTestGeneration/ProgramModification.cs [hunk 1]
```diff
@@ -136,27 +136,25 @@ private enum Status { Success, Failure, Untested }
...
           Console.WriteLine(
-            $"// No test can be generated for {uniqueId} " +
-            "because the verifier suceeded.");
-        } else if (log.Contains("MODEL")) {
-          Console.WriteLine(
-            $"// No test can be generated for {uniqueId} " +
-            "because there is no enhanced error trace.");
-        } else if (log.Contains("anon0")) {
+            $"// No test is generated for {uniqueId} " +
+            "because the verifier succeeded. This can be caused by dead code.");
```
[the remark this answers is not in the document]
Comment (Dargones) [reply Source/DafnyTestGeneration/ProgramModification.cs]: This makes sense - paraphrased this to "because the verifier proved that no inputs could cause this block to be visited"

File: Source/DafnyTestGeneration/Main.cs [hunk 2]
```diff
@@ -77,6 +79,22 @@ public static class Main {
         .ToList().ConvertAll(tuple => tuple.Item2);
       DafnyOptions.O.PrintInstrumented = oldPrintInstrumented;
 
+      if (DafnyOptions.O.TestGenOptions.TargetMethod != null) {
```
[the remark this answers is not in the document]
Comment (Dargones) [reply Source/DafnyTestGeneration/Main.cs]: Done!

**Comment by Dargones** [comment]
Thank you, @atomb. I noticed that the recent changes with DafnyOptions broke a few things, so I added minor fixes for these to this PR as well (two of the last three items on the list of changes above).

**Comment by Dargones** [comment]
Made a small change (commit above) to align this PR with the one I am preparing next but otherwise everything should be the same.

**Review by atomb — APPROVED** [review]
